src/main/java/me/jellysquid/mods/sodium/client/model/quad/ModelQuadView.java

```diff
@@ -58,6 +62,91 @@ public interface ModelQuadView {
      */
     Direction getLightFace();
 
+    /**
+     * @return The x coordinate of the unit normal vector
+     */
+    int getGFNINormX();
+
+    /**
+     * @return The y coordinate of the unit normal vector
+     */
+    int getGFNINormY();
+
+    /**
+     * @return The z coordinate of the unit normal vector
+     */
+    int getGFNINormZ();
+
+    /**
+     * Sets an accurate normal vector for this quad. This is used for GFNI.
+     * 
+     * @param x The normal's x component
+     * @param y The normal's y component
+     * @param z The normal's z component
+     */
+    void setGFNINormal(int x, int y, int z);
+
+    default int calculateNormals(boolean calculateUnitNormal) {
+        final float x0 = getX(0);
+        final float y0 = getY(0);
+        final float z0 = getZ(0);
+
+        final float x1 = getX(1);
+        final float y1 = getY(1);
+        final float z1 = getZ(1);
+
+        final float x2 = getX(2);
+        final float y2 = getY(2);
+        final float z2 = getZ(2);
+
+        final float x3 = getX(3);
+        final float y3 = getY(3);
+        final float z3 = getZ(3);
+
+        final float dx0 = x2 - x0;
+        final float dy0 = y2 - y0;
+        final float dz0 = z2 - z0;
+        final float dx1 = x3 - x1;
+        final float dy1 = y3 - y1;
+        final float dz1 = z3 - z1;
+
+        float normX = dy0 * dz1 - dz0 * dy1;
+        float normY = dz0 * dx1 - dx0 * dz1;
+        float normZ = dx0 * dy1 - dy0 * dx1;
+
+        int packedNormal = -1;
+        if (calculateUnitNormal) {
+            // normalize by length for the packed normal
+            float length = (float) Math.sqrt(normX * normX + normY * normY + normZ * normZ);
+            if (length != 0.0 && length != 1.0) {
+                normX /= length;
+                normY /= length;
+                normZ /= length;
+            }
+
+            packedNormal = NormI8.pack(normX, normY, normZ);
+        }
+
+        // normalize onto the surface of a cube by dividing by the length of the longest
+        // component
+        float infNormLength = Math.max(Math.abs(normX), Math.max(Math.abs(normY), Math.abs(normZ)));
+        if (infNormLength != 0 && infNormLength != 1) {
```

Comment:
you are still at risk of dividing by "almost zero" which will create grossly inaccurate results

Reply:
This part is not Sodium's code but where I make my own (heavily quantized) normal for GFNI triggering and translucency sorting. Since I'm projecting onto the surface of a unit cube anyways, this is probably fine. Unless it generates non-unit length normals it doesn't really matter if the result is weird if the input is weird.